README.rst
Servo Brick
===========

This repository contains the firmware source code and the hardware design
files. The documentation generator configs can be found at
https://github.com/Tinkerforge/generators

Repository Content
------------------

software/:
 * examples/: Examples for all supported languages
 * build/: Compiled files
 * src/: Source code of firmware
 * Makefile: Makefile to build project

hardware/:
 * Contains KiCad project files and additionally schematics as PDF

datasheets/:
 * Contains datasheets for sensors and complex ICs that are used

Hardware
--------

The hardware is designed with the open source EDA Suite KiCad
(http://www.kicad-pcb.org). Before you are able to open the files,
you have to install the Tinkerforge kicad-libraries
(https://github.com/Tinkerforge/kicad-libraries). You can either clone
them directly in hardware/ or clone them in a separate folder and
symlink them into hardware/
(ln -s kicad_path/kicad-libraries project_path/hardware). After that you
can open the .pro file in hardware/ with KiCad and from there view and
modify the schematics and the PCB layout.

Software
--------

If you want to do your own Brick/Bricklet firmware development we highly
recommend that you use our build environment setup script and read the
tutorial (https://www.tinkerforge.com/en/doc/Tutorials/Tutorial_Build_Environment/Tutorial.html).

To compile the C code we recommend you to install the newest GNU Arm Embedded
Toolchain (https://launchpad.net/gcc-arm-embedded/+download).
You also need to install bricklib (https://github.com/Tinkerforge/bricklib).

You can either clone it directly in software/src/ or clone it in a
separate folder and symlink it into software/src/
(ln -s bricklib_path/bricklib project_path/software/src/). Finally make sure to
have CMake installed (http://www.cmake.org/cmake/resources/software.html).

After that you can build the firmware by invoking make in software/.
The firmware (.bin) can then be found in software/build/ and uploaded
with brickv (click button "Flashing" on start screen).


software/changelog
2011-11-24: 1.0.0 (cc2f075)
- Initial version

2012-02-15: 1.1.0 (b80a126)
- Add support for tick types

2012-03-05: 1.1.1 (0b80822)
- Fix bug with timeouts on Windows (through bricklib)

2012-03-11: 1.1.2 (a7235ec)
- Fix bug with ADC calibration readout (through bricklib)

2012-06-30: 1.1.3 (91bfc80)
- Add get_chip_temperature and reset function

2012-07-02: 1.1.4 (c1bcd00)
- Add support for minimum period of 1us and add USB hotplug support

2012-10-26: 1.1.5 (bbcad00)
- Change Bricklet Context size from 250 to 256 (through bricklib)

2013-01-17: 2.0.0 (3c4227a)
- Support for Protocol 2.0

2013-03-27: 2.0.1 (2e5ca12)
- Add enable/disable API for reached callbacks

2013-04-11: 2.0.2 (5a58710)
- Add watchdog timer (restart after 16s of "deadlock")

2013-06-17: 2.0.3 (9eb0db7)
- Add missing defines for enable/disable API for reached callbacks

2013-06-27: 2.0.4 (d53add1)
- Use tick_task to detect USB connect/disconnect instead of interrupt

2014-07-07: 2.0.5 (77e3ed8)
- Remove USB hotplug support (through bricklib)

2014-11-25: 2.3.0 (1659022)
- Add support for new SPI protocol

2016-08-29: 2.3.1 (fd8c76a)
- Make sure to use all chunks when writing Bricklet plugin (through bricklib)

2017-05-11: 2.3.2 (5903619)
- Add support for Co-MCU Bricklets

2017-06-28: 2.3.3 (e737cdf)
- Fix incompatibility between Co-MCU support and LED Strip Bricklet (through bricklib)

2017-12-13: 2.3.4 (7e884e7)
- Make sure Brick does not restart on suspend/resume (through bricklib)
- Refactor initial enumeration handling, fix double-enumeration bugs (through bricklib)

2018-01-11: 2.3.5 (83558bd)
- Fix message duplication bug (through bricklib)

2018-04-27: 2.3.6 (15993f3)
- Fix potential deadlock between USB stack and Co-MCU enumeration (through bricklib)

2018-07-31: 2.3.7 (28e9cd9)
- Add support for Isolator Bricklet (through bricklib)
- Improve Co-MCU Bricklet error handling (through bricklib)

2019-02-27: 2.3.8 (c7067a8)
- Changes through bricklib:
  - Fix Bricklet port miss-identification
  - Increase SPI stack timeout (for RED Brick)
  - Fix bugs in set/get_spitfp_* functions

2019-11-22: 2.3.9 (ad70b6a)
- Make sure that Servo Brick output goes low if pulse width and position is set to 0


software/src/communication.h
/* servo-brick
 * Copyright (C) 2010-2013 Olaf Lüke <olaf@tinkerforge.com>
 *
 * communication.h: Implementation of Servo-Brick specific messages
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */


#ifndef COMMUNICATION_H
#define COMMUNICATION_H

#include <stdint.h>
#include <stdbool.h>

#include "bricklib/com/com_common.h"

#define FID_ENABLE 1
#define FID_DISABLE 2
#define FID_IS_ENABLED 3
#define FID_SET_POSITION 4
#define FID_GET_POSITION 5
#define FID_GET_CURRENT_POSITION 6
#define FID_SET_VELOCITY 7
#define FID_GET_VELOCITY 8
#define FID_GET_CURRENT_VELOCITY 9
#define FID_SET_ACCELERATION 10
#define FID_GET_ACCELERATION 11
#define FID_SET_OUTPUT_VOLTAGE 12
#define FID_GET_OUTPUT_VOLTAGE 13
#define FID_SET_PULSE_WIDTH 14
#define FID_GET_PULSE_WIDTH 15
#define FID_SET_DEGREE 16
#define FID_GET_DEGREE 17
#define FID_SET_PERIOD 18
#define FID_GET_PERIOD 19
#define FID_GET_SERVO_CURRENT 20
#define FID_GET_OVERALL_CURRENT 21
#define FID_GET_STACK_INPUT_VOLTAGE 22
#define FID_GET_EXTERNAL_INPUT_VOLTAGE 23
#define FID_SET_MINIMUM_VOLTAGE 24
#define FID_GET_MINIMUM_VOLTAGE 25
#define FID_UNDER_VOLTAGE 26
#define FID_POSITION_REACHED 27
#define FID_VELOCITY_REACHED 28
#define FID_ENABLE_POSITION_REACHED_CALLBACK 29
#define FID_DISABLE_POSITION_REACHED_CALLBACK 30
#define FID_IS_POSITION_REACHED_CALLBACK_ENABLED 31
#define FID_ENABLE_VELOCITY_REACHED_CALLBACK 32
#define FID_DISABLE_VELOCITY_REACHED_CALLBACK 33
#define FID_IS_VELOCITY_REACHED_CALLBACK_ENABLED 34

#define COM_MESSAGE_USER_LAST_FID 34

#define COM_MESSAGES_USER \
	{FID_ENABLE, (message_handler_func_t)enable}, \
	{FID_DISABLE, (message_handler_func_t)disable}, \
	{FID_IS_ENABLED, (message_handler_func_t)is_enabled}, \
	{FID_SET_POSITION, (message_handler_func_t)set_position}, \
	{FID_GET_POSITION, (message_handler_func_t)get_position}, \
	{FID_GET_CURRENT_POSITION, (message_handler_func_t)get_current_position}, \
	{FID_SET_VELOCITY, (message_handler_func_t)set_velocity}, \
	{FID_GET_VELOCITY, (message_handler_func_t)get_velocity}, \
	{FID_GET_CURRENT_VELOCITY, (message_handler_func_t)get_current_velocity}, \
	{FID_SET_ACCELERATION, (message_handler_func_t)set_acceleration}, \
	{FID_GET_ACCELERATION, (message_handler_func_t)get_acceleration}, \
	{FID_SET_OUTPUT_VOLTAGE, (message_handler_func_t)set_output_voltage}, \
	{FID_GET_OUTPUT_VOLTAGE, (message_handler_func_t)get_output_voltage}, \
	{FID_SET_PULSE_WIDTH, (message_handler_func_t)set_pulse_width}, \
	{FID_GET_PULSE_WIDTH, (message_handler_func_t)get_pulse_width}, \
	{FID_SET_DEGREE, (message_handler_func_t)set_degree}, \
	{FID_GET_DEGREE, (message_handler_func_t)get_degree}, \
	{FID_SET_PERIOD, (message_handler_func_t)set_period}, \
	{FID_GET_PERIOD, (message_handler_func_t)get_period}, \
	{FID_GET_SERVO_CURRENT, (message_handler_func_t)get_servo_current}, \
	{FID_GET_OVERALL_CURRENT, (message_handler_func_t)get_overall_current}, \
	{FID_GET_STACK_INPUT_VOLTAGE, (message_handler_func_t)get_stack_input_voltage}, \
	{FID_GET_EXTERNAL_INPUT_VOLTAGE, (message_handler_func_t)get_external_input_voltage}, \
	{FID_SET_MINIMUM_VOLTAGE, (message_handler_func_t)set_minimum_voltage}, \
    {FID_GET_MINIMUM_VOLTAGE, (message_handler_func_t)get_minimum_voltage}, \
    {FID_UNDER_VOLTAGE, (message_handler_func_t)NULL}, \
    {FID_POSITION_REACHED, (message_handler_func_t)NULL}, \
    {FID_VELOCITY_REACHED, (message_handler_func_t)NULL}, \
    {FID_ENABLE_POSITION_REACHED_CALLBACK, (message_handler_func_t)enable_position_reached_callback}, \
    {FID_DISABLE_POSITION_REACHED_CALLBACK, (message_handler_func_t)disable_position_reached_callback}, \
    {FID_IS_POSITION_REACHED_CALLBACK_ENABLED, (message_handler_func_t)is_position_reached_callback_enabled}, \
    {FID_ENABLE_VELOCITY_REACHED_CALLBACK, (message_handler_func_t)enable_velocity_reached_callback}, \
    {FID_DISABLE_VELOCITY_REACHED_CALLBACK, (message_handler_func_t)disable_velocity_reached_callback}, \
    {FID_IS_VELOCITY_REACHED_CALLBACK_ENABLED, (message_handler_func_t)is_velocity_reached_callback_enabled},

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) Enable;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) Disable;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) IsEnabled;

typedef struct {
	MessageHeader header;
	bool enabled;
} __attribute__((__packed__)) IsEnabledReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
	int16_t position;
} __attribute__((__packed__)) SetPosition;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetPosition;

typedef struct {
	MessageHeader header;
	int16_t position;
} __attribute__((__packed__)) GetPositionReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetCurrentPosition;

typedef struct {
	MessageHeader header;
	int16_t position;
} __attribute__((__packed__)) GetCurrentPositionReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
	uint16_t velocity;
} __attribute__((__packed__)) SetVelocity;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetVelocity;

typedef struct {
	MessageHeader header;
	uint16_t velocity;
} __attribute__((__packed__)) GetVelocityReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetCurrentVelocity;

typedef struct {
	MessageHeader header;
	uint16_t velocity;
} __attribute__((__packed__)) GetCurrentVelocityReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
	uint16_t acceleration;
} __attribute__((__packed__)) SetAcceleration;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetAcceleration;

typedef struct {
	MessageHeader header;
	uint16_t acceleration;
} __attribute__((__packed__)) GetAccelerationReturn;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) SetOutputVoltage;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetOutputVoltage;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) GetOutputVoltageReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
	uint16_t min_pulse_width;
	uint16_t max_pulse_width;
} __attribute__((__packed__)) SetPulseWidth;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetPulseWidth;

typedef struct {
	MessageHeader header;
	uint16_t min_pulse_width;
	uint16_t max_pulse_width;
} __attribute__((__packed__)) GetPulseWidthReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
	int16_t min_degree;
	int16_t max_degree;
} __attribute__((__packed__)) SetDegree;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetDegree;

typedef struct {
	MessageHeader header;
	int16_t min_degree;
	int16_t max_degree;
} __attribute__((__packed__)) GetDegreeReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
	uint16_t period;
} __attribute__((__packed__)) SetPeriod;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetPeriod;

typedef struct {
	MessageHeader header;
	uint16_t period;
} __attribute__((__packed__)) GetPeriodReturn;

typedef struct {
	MessageHeader header;
	uint8_t servo;
} __attribute__((__packed__)) GetServoCurrent;

typedef struct {
	MessageHeader header;
	uint16_t current;
} __attribute__((__packed__)) GetServoCurrentReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetOverallCurrent;

typedef struct {
	MessageHeader header;
	uint16_t current;
} __attribute__((__packed__)) GetOverallCurrentReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetStackInputVoltage;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) GetStackInputVoltageReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetExternalInputVoltage;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) GetExternalInputVoltageReturn;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) SetMinimumVoltage;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) GetMinimumVoltage;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) GetMinimumVoltageReturn;

typedef struct {
	MessageHeader header;
	uint16_t voltage;
} __attribute__((__packed__)) UnderVoltageCallback;

typedef struct {
	MessageHeader header;
	uint8_t servo;
	int16_t position;
} __attribute__((__packed__)) PositionReachedCallback;

typedef struct {
	MessageHeader header;
	uint8_t servo;
	int16_t velocity;
} __attribute__((__packed__)) VelocityReachedCallback;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) EnableVelocityReachedCallback;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) DisableVelocityReachedCallback;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) IsVelocityReachedCallbackEnabled;

typedef struct {
	MessageHeader header;
	bool enabled;
} __attribute__((__packed__)) IsVelocityReachedCallbackEnabledReturn;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) EnablePositionReachedCallback;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) DisablePositionReachedCallback;

typedef struct {
	MessageHeader header;
} __attribute__((__packed__)) IsPositionReachedCallbackEnabled;

typedef struct {
	MessageHeader header;
	bool enabled;
} __attribute__((__packed__)) IsPositionReachedCallbackEnabledReturn;

void enable(const ComType com, const Enable *data);
void disable(const ComType com, const Disable *data);
void is_enabled(const ComType com, const IsEnabled *data);
void set_position(const ComType com, const SetPosition *data);
void get_position(const ComType com, const GetPosition *data);
void get_current_position(const ComType com, const GetCurrentPosition *data);
void set_velocity(const ComType com, const SetVelocity *data);
void get_velocity(const ComType com, const GetVelocity *data);
void get_current_velocity(const ComType com, const GetCurrentVelocity *data);
void set_acceleration(const ComType com, const SetAcceleration *data);
void get_acceleration(const ComType com, const GetAcceleration *data);
void set_output_voltage(const ComType com, const SetOutputVoltage *data);
void get_output_voltage(const ComType com, const GetOutputVoltage *data);
void set_pulse_width(const ComType com, const SetPulseWidth *data);
void get_pulse_width(const ComType com, const GetPulseWidth *data);
void set_degree(const ComType com, const SetDegree *data);
void get_degree(const ComType com, const GetDegree *data);
void set_period(const ComType com, const SetPeriod *data);
void get_period(const ComType com, const GetPeriod *data);
void get_servo_current(const ComType com, const GetServoCurrent *data);
void get_overall_current(const ComType com, const GetOverallCurrent *data);
void get_stack_input_voltage(const ComType com, const GetStackInputVoltage *data);
void get_external_input_voltage(const ComType com, const GetExternalInputVoltage *data);
void set_minimum_voltage(const ComType com, const SetMinimumVoltage *data);
void get_minimum_voltage(const ComType com, const GetMinimumVoltage *data);
void enable_velocity_reached_callback(const ComType com, const EnableVelocityReachedCallback *data);
void disable_velocity_reached_callback(const ComType com, const DisableVelocityReachedCallback *data);
void is_velocity_reached_callback_enabled(const ComType com, const IsVelocityReachedCallbackEnabled *data);
void enable_position_reached_callback(const ComType com, const EnablePositionReachedCallback *data);
void disable_position_reached_callback(const ComType com, const DisablePositionReachedCallback *data);
void is_position_reached_callback_enabled(const ComType com, const IsPositionReachedCallbackEnabled *data);

#endif
